File: src/tools/comments.ts
import { McpServer } from "@mcp/server";
import { z } from "zod";
import type { LeantimeClient } from "../leantime-client.ts";
import { markdownToHtml } from "../markdown.ts";
import {
  CONFIRM_PARAM_DOC,
  checkDestructiveConfirmation,
  errorResult,
  isLeantimeError,
  MARKDOWN_HINT,
} from "./shared.ts";

export function registerCommentTools(server: McpServer, client: LeantimeClient) {
  server.tool(
    "leantime_list_comments",
    "List the discussion comments of a ticket",
    {
      ticketId: z.string().describe("The ticket ID"),
    },
    async ({ ticketId }) => {
      try {
        // Leantime expects `entityId` (not moduleId) — param names must match exactly.
        const result = await client.call<unknown[]>("comments.getComments", {
          module: "ticket",
          entityId: ticketId,
        });
        return {
          content: [{ type: "text", text: JSON.stringify(result ?? [], null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_add_comment",
    `Add a comment to a ticket's discussion. The text is ${MARKDOWN_HINT}.`,
    {
      ticketId: z.string().describe("The ticket ID"),
      text: z.string().describe(`Comment body in ${MARKDOWN_HINT}`),
      father: z.number().optional().describe(
        "Parent comment ID for a reply (omit or 0 for a top-level comment)",
      ),
    },
    async ({ ticketId, text, father }) => {
      try {
        // addComment requires the full `entity` object — rebuild it from the ticket.
        const ticket = await client.call<Record<string, unknown>>("tickets.getTicket", {
          id: ticketId,
        });
        if (typeof ticket === "boolean" || isLeantimeError(ticket)) {
          return errorResult(`Ticket ${ticketId} not found.`);
        }

        const html = markdownToHtml(text);
        try {
          await client.call<boolean>("comments.addComment", {
            values: { text: html, father: father ?? 0 },
            module: "ticket",
            entityId: ticketId,
            entity: {
              id: ticketId,
              type: ticket.type ?? "task",
              headline: ticket.headline,
            },
          });
        } catch (e) {
          // Leantime v3.7.3 bug: the comment row is inserted, then the
          // notification build crashes on entity property access via JSON-RPC.
          // Verify the comment actually landed before surfacing an error.
          const comments = await client.call<{ text?: string }[]>("comments.getComments", {
            module: "ticket",
            entityId: ticketId,
          });
          const landed = (comments ?? []).some((c) => c.text === html);
          if (!landed) throw e;
        }
        return {
          content: [{ type: "text", text: JSON.stringify({ ok: true, ticketId }, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_update_comment",
    `Edit an existing comment. The text is ${MARKDOWN_HINT}.`,
    {
      commentId: z.string().describe("The comment ID"),
      text: z.string().describe(`New comment body in ${MARKDOWN_HINT}`),
    },
    async ({ commentId, text }) => {
      try {
        const result = await client.call<boolean>("comments.editComment", {
          values: { text: markdownToHtml(text) },
          id: commentId,
        });
        return {
          content: [{ type: "text", text: JSON.stringify({ ok: result === true, commentId }, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_delete_comment",
    "Delete a comment. Destructive: requires explicit user approval (confirm: true) " +
      "unless LEANTIME_MCP_DESTRUCTIVE_POLICY is set otherwise.",
    {
      commentId: z.string().describe("The comment ID"),
      confirm: z.boolean().optional().describe(CONFIRM_PARAM_DOC),
    },
    async ({ commentId, confirm }) => {
      const check = checkDestructiveConfirmation(confirm, "comment");
      if (!check.allowed) return errorResult(check.message);
      try {
        const result = await client.call<boolean>("comments.deleteComment", {
          commentId,
        });
        return {
          content: [{
            type: "text",
            text: JSON.stringify({ deleted: result === true, commentId }, null, 2),
          }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );
}

File: src/tools/comments.rs
use std::future::Future;
use std::pin::Pin;

use serde_json::{json, Value};

use crate::markdown::markdown_to_html;

use super::shared::*;
use super::{error_result, ok_result, ClientRef, Tool, ToolAnnotations};

fn h_list_comments(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let tid = a.get("ticketId").and_then(|v| v.as_str()).unwrap_or("");
        match c
            .call(
                "comments.getComments",
                json!({"module": "ticket", "entityId": tid}),
            )
            .await
        {
            Ok(r) => ok_result(&json!(r.as_array().cloned().unwrap_or_default())),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_add_comment(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let tid = a.get("ticketId").and_then(|v| v.as_str()).unwrap_or("");
        let text = a.get("text").and_then(|v| v.as_str()).unwrap_or("");
        let father = a.get("father").and_then(|v| v.as_i64()).unwrap_or(0);
        let html = markdown_to_html(text);
        // addComment requires the full `entity` object — rebuild it from the ticket.
        let ticket = match c.call("tickets.getTicket", json!({"id": tid})).await {
            Ok(t) => t,
            Err(e) => return error_result(&e.to_string()),
        };
        if ticket.is_boolean() || is_leantime_error(&ticket) {
            return error_result(&format!("Ticket {} not found.", tid));
        }
        let add = c.call("comments.addComment", json!({
            "values": {"text": html, "father": father}, "module": "ticket", "entityId": tid,
            "entity": {"id": tid, "type": ticket.get("type").cloned().unwrap_or(json!("task")), "headline": ticket.get("headline")}
        })).await;
        if add.is_err() {
            // Leantime v3.7.3 bug: the comment row is inserted, then the
            // notification build crashes on entity property access via JSON-RPC.
            // Verify the comment actually landed before surfacing an error.
            let landed = c
                .call(
                    "comments.getComments",
                    json!({"module": "ticket", "entityId": tid}),
                )
                .await
                .ok()
                .and_then(|comments| comments.as_array().cloned())
                .map(|arr| {
                    arr.iter()
                        .any(|cm| cm.get("text").and_then(|t| t.as_str()) == Some(html.as_str()))
                })
                .unwrap_or(false);
            if !landed {
                return error_result(&add.err().unwrap().to_string());
            }
        }
        ok_result(&json!({ "ok": true, "ticketId": tid }))
    })
}

fn h_update_comment(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let id = a.get("commentId").and_then(|v| v.as_str()).unwrap_or("");
        let text = a.get("text").and_then(|v| v.as_str()).unwrap_or("");
        match c
            .call(
                "comments.editComment",
                json!({"values": {"text": markdown_to_html(text)}, "id": id}),
            )
            .await
        {
            Ok(r) => ok_result(&json!({ "ok": r == json!(true), "commentId": id })),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_delete_comment(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        if let Err(e) = check_destructive(a.get("confirm").and_then(|v| v.as_bool()), "comment") {
            return error_result(&e);
        }
        let mut c = cl.lock().await;
        let id = a.get("commentId").and_then(|v| v.as_str()).unwrap_or("");
        match c
            .call("comments.deleteComment", json!({ "commentId": id }))
            .await
        {
            Ok(r) => ok_result(&json!({ "deleted": r == json!(true), "commentId": id })),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

pub(super) fn tools() -> Vec<Tool> {
    let md = MARKDOWN_HINT;
    vec![
        tool("leantime_list_comments", "List the discussion comments of a ticket",
            vec![rs("ticketId", "The ticket ID")], vec!["ticketId"], Box::new(h_list_comments)),
        tool_with_annotations("leantime_add_comment", format!("Add a comment to a ticket's discussion. The text is {}.", md),
            vec![rs("ticketId", "The ticket ID"), rs("text", format!("Comment body in {}", md)), ("father".to_string(), json!({"type": "number", "description": "Parent comment ID for a reply (omit or 0 for a top-level comment)", "optional": true}))],
            vec!["ticketId", "text"], Box::new(h_add_comment), ToolAnnotations::write()),
        tool_with_annotations("leantime_update_comment", format!("Edit an existing comment. The text is {}.", md),
            vec![rs("commentId", "The comment ID"), rs("text", format!("New comment body in {}", md))],
            vec!["commentId", "text"], Box::new(h_update_comment), ToolAnnotations::write()),
        tool_with_annotations("leantime_delete_comment", "Delete a comment. Destructive: requires explicit user approval (confirm: true) unless LEANTIME_MCP_DESTRUCTIVE_POLICY is set otherwise.",
            vec![rs("commentId", "The comment ID"), ob("confirm", "MUST be true to actually delete (ask the user for explicit approval first)")],
            vec!["commentId"], Box::new(h_delete_comment), ToolAnnotations::destructive()),
    ]
}
